### ISSUE TITLE
Bug: Parsers fails to cache ts.program

### ISSUE THREAD

**Issue by deen13** [typescript-eslint/typescript-eslint#9350]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

Linting our project currently takes an excessively long time and fails with `0:0  error  Parsing error: Maximum call stack size exceeded` once it reaches the memory limit. This issue does not occur when either using the new experimental project service feature or enabling the `allowAutomaticSingleRunInference` flag. 

### Problem Analysis

The root cause of the issue has been isolated to [this line of code](https://github.com/typescript-eslint/typescript-eslint/blob/ced65ed5127dad2baa28c8177084d1828cb534cf/packages/typescript-estree/src/create-program/getWatchProgramsForProjects.ts#L366) in [getWatchProgramsForProjects.ts](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-estree/src/create-program/getWatchProgramsForProjects.ts). It appears that the `getWatchProgramsForProjects.ts` attempts to cache the program created for the first file and reuse it for subsequent files. However, on my machine, `ts.createWatchProgram(...)` never returns, causing subsequent files to encounter an empty cache scenario and create new programs. This ultimately leads to memory exhaustion.

### Demonstration

To illustrate this behavior, I have added logging around `ts.createWatchProgram(watchCompilerHost)`:

```typescript
log("Calling ts.createWatchProgram...");
const watchProgram = ts.createWatchProgram(watchCompilerHost);
log("Returning result of ts.createWatchProgram...");
return watchProgram;
```

This produces the following log entries for each file until the memory limit is reached and the error occurs:

```
typescript-eslint:typescript-estree:createWatchProgram Creating watch program for /Users/deen/Coding/integreat-cms/tsconfig.json. +0ms
typescript-eslint:typescript-estree:createWatchProgram Calling ts.createWatchProgram... +0ms
typescript-eslint:typescript-estree:createWatchProgram File did not belong to any existing programs, moving to create/update. /users/deen/coding/integreat-cms/integreat_cms/static/src/js/chat/send-chat-message.ts +604ms
typescript-eslint:typescript-estree:createWatchProgram Creating watch program for /Users/deen/Coding/integreat-cms/tsconfig.json. +0ms
typescript-eslint:typescript-estree:createWatchProgram Calling ts.createWatchProgram... +0ms
typescript-eslint:typescript-estree:createWatchProgram File did not belong to any existing programs, moving to create/update. /users/deen/coding/integreat-cms/integreat_cms/static/src/js/collapsible-boxes.ts +612ms
typescript-eslint:typescript-estree:createWatchProgram Creating watch program for /Users/deen/Coding/integreat-cms/tsconfig.json. +0ms
typescript-eslint:typescript-estree:createWatchProgram Calling ts.createWatchProgram... +0ms
typescript-eslint:typescript-estree:createWatchProgram File did not belong to any existing programs, moving to create/update. /users/deen/coding/integreat-cms/integreat_cms/static/src/js/conditional-fields.ts +619ms
typescript-eslint:typescript-estree:createWatchProgram Creating watch program for /Users/deen/Coding/integreat-cms/tsconfig.json. +0ms
```

### Configuration

This is the relevant part of our eslint configuration:

```
parser: "@typescript-eslint/parser",
parserOptions: {
    ecmaVersion: 6,
    sourceType: "module",
    project: "./tsconfig.json",
    ecmaFeatures: {
        jsx: true,
    },
},
```

### Reproduction Repository Link

https://github.com/digitalfabrik/integreat-cms

### Repro Steps

1) clone the repo
2) `npm install`
3) `DEBUG=typescript-eslint:* ./node_modules/.bin/eslint .`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `7.12.0` |
| `@typescript-eslint/parser`            | `7.12.0` |
| `@typescript-eslint/scope-manager`     | `7.12.0` |
| `@typescript-eslint/typescript-estree` | `7.12.0` |
| `@typescript-eslint/type-utils`        | `7.12.0` |
| `@typescript-eslint/utils`             | `7.12.0` |
| `TypeScript`                           | `5.2.2` |
| `ESLint`                               | `8.57.0` |
| `node`                                 | `v21.7.0` |

**Comment by bradzacher** [comment]
> However, on my machine, ts.createWatchProgram(...) never returns

This is impossible - if that function call never returns then the entire lint run would physically stop. ESLint does not do async/multi-threaded linting and only does things serially.

So if that call never returns the code would just hang infinitely.
So there is likely another cause.

**Comment by bradzacher** [comment]
First thing I can see - your repo doesn't specify an explicit dependency on our `eslintrc` tooling.
You've specified `"typescript-eslint": "^7.11.0"` - but that package is for flat configs and introduces a transient dependency on our packages.

This creates a bad install setup wherein the top-level version of our packages is the old v6 version and that's the version ESLint uses. I ran `npm i` on your repo and when I ran `npx eslint .` I got errors about an unsupported TS version.

```
$ npm ls @typescript-eslint/parser
integreat-cms-develop@ /Users/bjz/temp/integreat-cms-develop
├─┬ eslint-config-preact@1.4.0
│ └─┬ eslint-plugin-jest@25.7.0
│   └─┬ @typescript-eslint/eslint-plugin@5.62.0
│     └── @typescript-eslint/parser@5.62.0
├─┬ eslint-plugin-jest@28.5.0
│ └─┬ @typescript-eslint/eslint-plugin@6.21.0
│   └── @typescript-eslint/parser@6.21.0 deduped
├─┬ prettier-eslint@16.3.0
│ └── @typescript-eslint/parser@6.21.0
└─┬ typescript-eslint@7.12.0
  ├─┬ @typescript-eslint/eslint-plugin@7.12.0
  │ └── @typescript-eslint/parser@7.12.0 deduped
  └── @typescript-eslint/parser@7.12.0
```

If you want to use flat configs - use `typescript-eslint`. If you aren't then use `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` directly.

**Comment by bradzacher** [comment]
I'm also unable to reproduce any OOMs or the like from your repo.
```
npx eslint .                    

/Users/bjz/temp/integreat-cms-develop/integreat_cms/static/src/js/analytics/hix-widget.ts
   13:12  warning  Unnecessary 'else' after 'return'  no-else-return
  130:12  warning  Unnecessary 'else' after 'return'  no-else-return

✖ 2 problems (0 errors, 2 warnings)
  0 errors and 2 warnings potentially fixable with the `--fix` option.
```

This works fine for me. Is there more to the repro than just running install + lint?

**Comment by deen13** [comment]
Hey @bradzacher,

Thanks for the fast reply! I’m also puzzled by this behavior. Unfortunately, the issue persists even when we directly depend on "@typescript-eslint/eslint-plugin" and "@typescript-eslint/parser".

It’s worth noting that this behavior seems environment-dependent. It works fine in our CircleCI workflow (and obviously on your machine), as you can see in the following logs.

```
typescript-eslint:typescript-estree:createWatchProgram File did not belong to any existing programs, moving to create/update. /home/circleci/project/integreat_cms/static/src/editor.ts +0ms
typescript-eslint:typescript-estree:createWatchProgram Creating watch program for /home/circleci/project/tsconfig.json. +0ms
typescript-eslint:typescript-estree:createWatchProgram Found program for file. /home/circleci/project/integreat_cms/static/src/editor.ts +2s
typescript-eslint:typescript-estree:createProjectProgram Creating project program for: /home/circleci/project/integreat_cms/static/src/editor.ts +0ms
typescript-eslint:parser:parser Resolved libs from program: [ 'es6' ] +0ms
```

Do you have any other ideas on what might be causing this? I've attached a screen recording that demonstrates the behavior on my MacBook for additional context.

https://github.com/typescript-eslint/typescript-eslint/assets/13850172/84937d33-7632-4b53-be2d-d8c3ab131711

**Comment by bradzacher** [comment]
There's likely something wrong with your environment in some way.
You only have a hundred or so files - so there should never be any chance of an OOM on that few files.

But I don't have a good answer for you without a solid reproduction case.
It's not a problem we've ever seen happen before.

**Comment by JoshuaKGoldberg** [comment]
I also can't reproduce this, running on https://github.com/digitalfabrik/integreat-cms/tree/c649006de559a03e8a3d4edf659b352318b40206. Great to see that you resolved it with the new project service though

Since we two of us can't repro it and the problem only reproduces in some specific environment settings, closing as non-actionable on our end (i.e. the problem is not know to be in typescript-eslint). But please do post back / file a new issue if you ever do get a reproduction. Thanks!
